Introduce the `Test.expectFailure` function

m-Peter opened the pull request:
Closes https://github.com/onflow/cadence-tools/issues/108

## Description

Introduces a new native function for allowing/expecting a test case to fail/panic, with a specified error message.
______

<!-- Complete: -->

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

m-Peter commented:
I will certainly add more tests, I just opened this up in order to bootstrap a conversation around the implementation details and desired functionality :pray:

turbolent commented on runtime/stdlib/test.go:
```diff
@@ -549,6 +562,76 @@ var testExpectFunction = interpreter.NewUnmeteredHostFunctionValue(
+	expectFailureFunctionType,
+	func(invocation interpreter.Invocation) interpreter.Value {
+		inter := invocation.Interpreter
+		functionValue := invocation.Arguments[0].(interpreter.FunctionValue)
```
```suggestion
		functionValue, ok := invocation.Arguments[0].(interpreter.FunctionValue)
		if !ok {
			panic(errors.NewUnreachableError())
		}
```

m-Peter replied:
Fixed in https://github.com/onflow/cadence/pull/2468/commits/136bc2a56bfa322ae2086f7bc6fa1027090276c2

m-Peter commented on runtime/stdlib/test.go:
```diff
@@ -549,6 +562,76 @@ var testExpectFunction = interpreter.NewUnmeteredHostFunctionValue(
 
+// Test.expectFailure function
+const expectFailureFunctionDocString = `
+Wraps a function call in a closure, and expects it to fail with a certain error message.
```
Fixed in https://github.com/onflow/cadence/pull/2468/commits/136bc2a56bfa322ae2086f7bc6fa1027090276c2

m-Peter commented on runtime/stdlib/test.go:
```diff
@@ -549,6 +562,76 @@ var testExpectFunction = interpreter.NewUnmeteredHostFunctionValue(
+			),
+		},
+		{
+			Label:      sema.ArgumentLabelNotRequired,
```
Fixed in https://github.com/onflow/cadence/pull/2468/commits/136bc2a56bfa322ae2086f7bc6fa1027090276c2
I changed it to `errorMessageSubstring`

turbolent replied:
That's a good name 👍 

Still need to remove the line here, and add the label in the calls of the tests.

```suggestion
```

m-Peter commented on runtime/stdlib/test.go:
```diff
@@ -549,6 +562,76 @@ var testExpectFunction = interpreter.NewUnmeteredHostFunctionValue(
+		_, _ = inter.InvokeExternally(
+			functionValue,
+			functionType,
+			[]interpreter.Value{},
```
Fixed in https://github.com/onflow/cadence/pull/2468/commits/136bc2a56bfa322ae2086f7bc6fa1027090276c2

m-Peter commented on runtime/stdlib/test.go:
```diff
@@ -549,6 +562,76 @@ var testExpectFunction = interpreter.NewUnmeteredHostFunctionValue(
+			Identifier: "functionWrapper",
+			TypeAnnotation: sema.NewTypeAnnotation(
+				&sema.FunctionType{
+					Parameters: []sema.Parameter{},
```
Fixed in https://github.com/onflow/cadence/pull/2468/commits/136bc2a56bfa322ae2086f7bc6fa1027090276c2

m-Peter commented:
@turbolent I have added more test cases, and fixed the scenario where a failure is expected, but there is none (https://github.com/onflow/cadence/pull/2468/commits/946ffae51fa713dda0dea8cb856ed6ce78e5d785#diff-4709f76365b48c2caca87b859afcc573afb62cb4e6186b9431d2b540972d402fR619)
However, I'm struggling with the test case here (https://github.com/onflow/cadence/pull/2468/commits/946ffae51fa713dda0dea8cb856ed6ce78e5d785#diff-74ad93a1435e307865ff8474567598f410879038788126ec263e27c9244edf98R1487). With a local build, the actual tests (in Cadence Testing Framework), works well. Do you have any clue?
P.S: This PR should be merged after https://github.com/onflow/cadence/pull/2403, to avoid huge conflicts :innocent:

m-Peter commented:
```cadence
import Test

pub struct Foo {
    priv let answer: UInt8
    
    init(answer: UInt8) {
        self.answer = answer
    }
    
    pub fun correctAnswer(_ input: UInt8): Bool {
        if self.answer != input {
            panic("wrong answer!")
        }
        return true
    }
}

pub fun testOne() {
    let foo = Foo(answer: 42)
    Test.expectFailure(fun(): Void {
        foo.correctAnswer(42)
    }, errorMessageSubstring: "wrong answer")
}

pub fun testTwo() {
    let foo = Foo(answer: 42)
    Test.expectFailure(fun(): Void {
        foo.correctAnswer(43)
    }, errorMessageSubstring: "wrong answer!")
}

pub fun testThree() {
    let foo = Foo(answer: 42)
    Test.expectFailure(fun(): Void {
        foo.correctAnswer(43)
    }, errorMessageSubstring: "what is wrong?!")
}
```

And the equivalent test execution results:

```bash
~/Dev/forks/flow-cli/cmd/flow/flow-x86_64-linux- test my_expect_failure_tests.cdc

Test results: "my_expect_failure_tests.cdc"
- FAIL: testOne
		Execution failed:
			error: Execution failed:
			error: Expected a failure, but found none.
			  --> 7465737400000000000000000000000000000000000000000000000000000000:14:8
			
			  --> 7465737400000000000000000000000000000000000000000000000000000000:14:8
			
- PASS: testTwo
- FAIL: testThree
		Execution failed:
			error: Expected error message to include: "what is wrong?!". Found: Execution failed:
			error: panic: wrong answer!
			  --> 7465737400000000000000000000000000000000000000000000000000000000:12:12
			
			  --> 7465737400000000000000000000000000000000000000000000000000000000:12:12
```

turbolent commented:
@m-Peter figured out what's wrong and also refactored the recover logic, but can't push the commits.

Could you please allow commits from maintainers by checking the "Allow edits from maintainers" box? (see https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)

m-Peter commented:
@turbolent It seems that it's not possible for organization repositories.. See https://github.com/orgs/community/discussions/5634 I am unable to find this option also..

m-Peter commented:
@turbolent I have sent you an invitation as a collaborator, you should be able to do so now :pray:

turbolent commented:
@m-Peter Thank you! Pushed up the fix. The issue was that the `panic` function was declared for the checker, but not for the interpreter

turbolent reviewed (approved):
Nice! Thanks for adding this

m-Peter commented:
> @m-Peter Thank you! Pushed up the fix. The issue was that the `panic` function was declared for the checker, but not for the interpreter

Oh, I see, great catch :100: :100: :100: 

I tried locally the newest changes, but it seems that they affect the displayed error messages. For example:

![Screenshot from 2023-04-27 12-22-25](https://user-images.githubusercontent.com/1778965/234832930-5fb076dc-76a8-482c-b76a-51e596dbc039.png)

I have added a commit to make them less verbose:

![Screenshot from 2023-04-27 12-32-17](https://user-images.githubusercontent.com/1778965/234833071-8bd8741a-91cc-4ad8-88bb-cd9dd201c57a.png)

What do you think about it? I can drop the changes, if you want.

turbolent reviewed (approved):
Looks better, nice!

turbolent commented:
@m-Peter #2403 is now merged, so there are some conflicts now. The additions of this PR basically have to be moved to `test_contract.go`

m-Peter commented:
@turbolent It was rather impossible to rebase on master, after https://github.com/onflow/cadence/pull/2403 was merged. So I had to reset the commits of this PR, manually apply the changes and then force-push. I have added you as co-author though :pray:

turbolent reviewed (approved):
Nice!
